velox/functions/sparksql/Split.h

```diff
@@ -0,0 +1,159 @@
+ */
+#pragma once
+
+#include "velox/functions/prestosql/Utf8Utils.h"
+
+namespace facebook::velox::functions::sparksql {
+
+/// split(string, delimiter[, limit]) -> array(varchar)
+///
+/// Splits string on delimiter and returns an array of size at most limit.
+/// delimiter is a string representing regular expression.
+/// limit is an integer which controls the number of times the regex is applied.
+/// By default, limit is -1.
+template <typename T>
+struct Split {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Array<Varchar>>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& delimiter) {
+    doCall(result, input, delimiter, INT32_MAX);
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Array<Varchar>>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& delimiter,
+      const arg_type<int32_t>& limit) {
+    doCall(result, input, delimiter, limit > 0 ? limit : INT32_MAX);
+  }
+
+ private:
+  void doCall(
+      ArrayWriter<Varchar>& result,
+      StringView input,
+      StringView delimiter,
+      int32_t limit) const {
+    if (delimiter.empty()) {
+      splitEmptyDelimiter(result, input, limit);
+    } else {
+      split(result, input, delimiter, limit);
+    }
+  }
+
+  // When pattern is empty, split each character out. Since Spark 3.4, when
+  // delimiter is empty, the result does not include an empty tail string, e.g.
+  // split('abc', '') outputs ["a", "b", "c"] instead of ["a", "b", "c", ""].
+  // The result does not include remaining string when limit is smaller than the
+  // string size, e.g. split('abc', '', 2) outputs ["a", "b"] instead of ["a",
+  // "bc"].
+  void splitEmptyDelimiter(
+      ArrayWriter<Varchar>& result,
+      StringView input,
+      int32_t limit) const {
+    if (input.size() == 0) {
+      result.add_item().setNoCopy(StringView());
+      return;
+    }
+
+    const size_t end = input.size();
+    const char* start = input.data();
+    size_t pos = 0;
+    int32_t count = 0;
+    while (pos < end && count < limit) {
+      auto charLength = tryGetCharLength(start + pos, end - pos);
+      VELOX_DCHECK_GT(charLength, 0);
+      result.add_item().setNoCopy(StringView(start + pos, charLength));
+      pos += charLength;
+      count += 1;
+    }
+  }
+
+  // Split with a non-empty delimiter. If limit > 0, The resulting array's
+  // length will not be more than limit and the resulting array's last entry
+  // will contain all input beyond the last matched regex. If limit <= 0,
+  // delimiter will be applied as many times as possible, and the resulting
+  // array can be of any size.
+  void split(
+      ArrayWriter<Varchar>& result,
+      StringView input,
+      StringView delim,
+      int32_t limit) const {
+    VELOX_USER_CHECK(!delim.empty(), "Non-empty delimiter is expected");
```

Comment:
VELOX_DCHECK, This logic is guaranteed by you.
```
spark-sql (default)> select split("a,b,c", "");
["a",",","b",",","c"]
Time taken: 0.093 seconds, Fetched 1 row(s)
```
The delimiter can be empty. Curious why this cannot be tested. Do I miss something?

Reply:
please  check `splitEmptyDelimiter` ~~

Reply:
So it should `VELOX_DCHECK`, user won't meet this exception